subprojects/core/src/integTest/groovy/org/gradle/api/invocation/GradleLifecycleAllprojectsIntegrationTest.groovy

```diff
@@ -0,0 +1,283 @@
+            }
+            include(":a")
+        """
+        file("a/build.gradle") << ""
+
+        when:
+        run "help", "-q"
+
+        then:
+        outputContains """
+lifecycle.allprojects for root
+lifecycle.beforeProject for root
+lifecycle.allprojects for a
+lifecycle.beforeProject for a
+"""
+    }
+
+    def 'lifecycle.allprojects is executed before project.#api'() {
+        given:
+        settingsFile << """
+            rootProject.name = 'root'
+
+            gradle.lifecycle.allprojects { project ->
+                project.ext {
+                    foo = "bar"
+                }
+            }
+            include(":a")
+            include(":b")
+        """
+
+        file("a/build.gradle") << ""
+        file("b/build.gradle") << ""
+
+        buildFile << """
+            $invocation { project ->
+                println "foo = \${project.foo} for \${project.name}"
+            }
+        """
+
+        when:
+        run "help", "-q"
+
+        then:
+        outputContains expectedOutput
+
+        where:
+        api              | invocation                 | expectedOutput
+        "allprojects"    | "allprojects"              | allprojectsExpectedOutput
+        "subprojects"    | "subprojects"              | subprojectsExpectedOutput
+        "project"        | "project(':a')"            | projectExpectedOutput
+        "getAllprojects" | "getAllprojects().forEach" | getAllprojectsExpectedOutput
+        "getSubprojects" | "getSubprojects().forEach" | getSubprojectsExpectedOutput
+    }
+
+    private static def allprojectsExpectedOutput = """
+foo = bar for root
+foo = bar for a
+foo = bar for b
+"""
+    private static def subprojectsExpectedOutput = """
+foo = bar for a
+foo = bar for b
+"""
+    private static def projectExpectedOutput = """
+foo = bar for a
+"""
+    private static def getAllprojectsExpectedOutput = """
+foo = bar for root
+foo = bar for a
+foo = bar for b
+"""
+    private static def getSubprojectsExpectedOutput = """
+foo = bar for a
+foo = bar for b
+"""
+
+    def 'lifecycle.allprojects is executed eagerly only if a mutable state of a project touched by using project.#api'() {
+        given:
+        settingsFile << """
+            rootProject.name = 'root'
+            gradle.lifecycle.allprojects { project ->
+                println "lifecycle.allprojects for \${project.name}"
+                project.ext {
+                    foo = "bar"
+                }
+            }
+            include(":a")
+            include(":b")
+        """
+
+        file("a/build.gradle") << "println 'a'"
+        file("b/build.gradle") << """
+            println 'b'
+            println "foo = \${foo} for b"
+        """
+
+        buildFile << """
+            println 'root'
+            $invocation { project ->
+                if (project.getName() == 'a') {
+                    println "Mutable state access for 'a': foo = \${project.foo}"
```

Comment:
💅  The mutable access happens inside the `println` statement, making it a bit harder to track.
What about having two separate statements for "before" and "after" mutable access, and a separate `def value = project.foo` to make that explicit?

Reply:
> having two separate statements for "before" and "after" mutable access

Can you show what you mean?

Reply:
```suggestion
                    println "Before mutable access for 'a'"
                    def value = project.foo
                    println "After mutable state access for 'a': foo = \$value"
```

Reply:
The new `allprojects` callback is then guaranteed to be executed between the println statements